crates/wasmparser/src/validator/operators.rs

```diff
@@ -1357,20 +1393,22 @@ where
         self.push_operand(ty)?;
         Ok(())
     }
-    fn visit_typed_select(&mut self, ty: ValType) -> Self::Output {
+    fn visit_typed_select(&mut self, mut ty: ValType) -> Self::Output {
         self.resources
             .check_value_type(ty, &self.features, self.offset)?;
         self.pop_operand(Some(ValType::I32))?;
         self.pop_operand(Some(ty))?;
         self.pop_operand(Some(ty))?;
+        self.resources.canonicalize_valtype(&mut ty);
         self.push_operand(ty)?;
         Ok(())
     }
     fn visit_local_get(&mut self, local_index: u32) -> Self::Output {
-        let ty = self.local(local_index)?;
+        let mut ty = self.local(local_index)?;
         if !self.local_inits[local_index as usize] {
             bail!(self.offset, "uninitialized local: {}", local_index);
         }
+        self.resources.canonicalize_valtype(&mut ty);
```

Comment:
Could this be avoided by canonicalizing locals when they're declared rather than on each instruction?